app/scripts/lib/EntryBridge.ts

```diff
@@ -0,0 +1,201 @@
+  const newEntriesMap = new Map(newEntries.map((e) => [getKey(e), e]));
+
+  newEntriesMap.forEach((newEntry, newKey) => {
+    const oldEntry = prevEntriesMap.get(newKey);
+    if (oldEntry) {
+      if (!areEqual(newEntry, oldEntry)) {
+        updated.push(newEntry);
+      }
+    } else {
+      added.push(newEntry);
+    }
+  });
+
+  prevEntriesMap.forEach((oldEntry, oldKey) => {
+    if (!newEntriesMap.has(oldKey)) {
+      deleted.push(oldEntry);
+    }
+  });
+
+  return { added, updated, deleted };
+}
+
+/**
+ * Applies the changes from a ChangeList to a given function.
+ *
+ * @param changeList - The list of changes.
+ * @param applyChange - The function to apply changes.
+ */
+function applyChangeList<Entry>(
+  changeList: ChangeList<Entry>,
+  applyChange: ApplyChangeCallback<Entry>,
+): void {
+  const { added, updated, deleted } = changeList;
+  added.forEach((entry) => applyChange('added', entry));
+  updated.forEach((entry) => applyChange('updated', entry));
+  deleted.forEach((entry) => applyChange('deleted', entry));
+}
+
+/**
+ * Configuration for a one-way bridge between entries in a source and entries
+ * in a destination.
+ *
+ * @param comparators - The object containing comparison functions.
+ * @param getSourceEntries - A function that returns the entries from the
+ * source.
+ * @param applyChangeToDestination - A function that applies a change to the
+ * destination.
+ */
+export type OneWayBridgeConfig<Entry> = {
+  comparators: EntryComparators<Entry>;
+  getSourceEntries: () => Entry[];
+  applyChangeToDestination: ApplyChangeCallback<Entry>;
+};
+
+/**
+ * A one-way bridge between the entries in a source and the entries in
+ * a destination.
+ */
+export class OneWayBridge<Entry> {
+  #comparators: EntryComparators<Entry>;
+
+  #getSourceEntries: () => Entry[];
+
+  #applyChangeToDestination: ApplyChangeCallback<Entry>;
+
+  #prevEntries: Entry[];
+
+  constructor({
+    comparators,
+    getSourceEntries,
+    applyChangeToDestination,
+  }: OneWayBridgeConfig<Entry>) {
+    this.#comparators = comparators;
+    this.#getSourceEntries = getSourceEntries;
+    this.#applyChangeToDestination = applyChangeToDestination;
+    this.#prevEntries = getSourceEntries();
+  }
+
+  public synchronize(): void {
+    const newEntries = this.#getSourceEntries();
+    const changeList = computeChangeList(
+      this.#prevEntries,
+      newEntries,
+      this.#comparators,
+    );
+    applyChangeList(changeList, this.#applyChangeToDestination);
+    this.#prevEntries = newEntries;
+  }
+}
+
+type Direction = 'L->R' | 'R->L';
+
+type TwoWayBridgeConfig<Entry> = {
+  comparators: EntryComparators<Entry>;
+  leftEntries: () => Entry[];
+  rightEntries: () => Entry[];
+  applyChangeToLeft: (type: ChangeType, entry: Entry) => void;
+  applyChangeToRight: (type: ChangeType, entry: Entry) => void;
+};
+
+// Manages a two-way bridge for entries, preventing recursive updates
+export class TwoWayBridge<Entry> {
```

Comment:
As mentioned in the other comment, I think we would simplify out logic if we define both the one way and two way bridges in a single abstract class and use a constructor boolean argument to request two way support or not.